File: main/store.js
const fs = require('fs');
const path = require('path');
const Store = require('electron-store');

const settingsStore = new Store({
  name: 'gh-stats-settings',
  defaults: {
    alwaysOnTop: false,
    startMinimised: false,
    windowBounds: null,
    lastFetchAt: null,
    autoFetchHours: 24,
    starHistoryLoaded: false,
    showDebugBar: false,
    timings: []
  }
});

const MAX_TIMINGS = 50;

function recordTiming(kind, ms, meta = {}) {
  const msVal = Math.round(Number(ms));
  if (!Number.isFinite(msVal) || msVal < 0) return;
  const list = settingsStore.get('timings', []);
  list.push({ kind, ms: msVal, at: Date.now(), ...meta });
  while (list.length > MAX_TIMINGS) list.shift();
  settingsStore.set('timings', list);
  return list[list.length - 1];
}

function avgMs(items) {
  if (!items.length) return null;
  return Math.round(items.reduce((sum, t) => sum + t.ms, 0) / items.length);
}

function getTimingStats() {
  const list = settingsStore.get('timings', []);
  const auth = list.filter((t) => t.kind === 'auth');
  const fetch = list.filter((t) => t.kind === 'fetch');
  const history = list.filter((t) => t.kind === 'history');
  const last = (items) => items[items.length - 1] || null;
  return {
    authAvgMs: avgMs(auth),
    fetchAvgMs: avgMs(fetch),
    historyAvgMs: avgMs(history),
    authCount: auth.length,
    fetchCount: fetch.length,
    historyCount: history.length,
    lastAuth: last(auth),
    lastFetch: last(fetch),
    lastHistory: last(history)
  };
}

function snapshotsDir(userDataPath) {
  const dir = path.join(userDataPath, 'snapshots');
  if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
  return dir;
}

function listSnapshotFiles(userDataPath) {
  const dir = snapshotsDir(userDataPath);
  return fs.readdirSync(dir)
    .filter((f) => f.endsWith('.json'))
    .sort();
}

function loadSnapshot(userDataPath, fileName) {
  const file = path.join(snapshotsDir(userDataPath), fileName);
  try {
    return JSON.parse(fs.readFileSync(file, 'utf8'));
  } catch {
    return null;
  }
}

function loadAllSnapshots(userDataPath) {
  return listSnapshotFiles(userDataPath)
    .map((f) => loadSnapshot(userDataPath, f))
    .filter(Boolean)
    .sort((a, b) => a.timestamp - b.timestamp);
}

function saveSnapshot(userDataPath, data) {
  const snapshot = {
    timestamp: data.timestamp || Date.now(),
    date: data.date || new Date().toISOString().slice(0, 10),
    repos: data.repos || [],
    totals: data.totals || { stars: 0, downloads: 0 }
  };
  const file = path.join(snapshotsDir(userDataPath), `${snapshot.date}.json`);
  fs.writeFileSync(file, JSON.stringify(snapshot, null, 2), 'utf8');
  settingsStore.set('lastFetchAt', snapshot.timestamp);
  return snapshot;
}

function getSettings() {
  return {
    alwaysOnTop: settingsStore.get('alwaysOnTop', false),
    startMinimised: settingsStore.get('startMinimised', false),
    lastFetchAt: settingsStore.get('lastFetchAt', null),
    autoFetchHours: settingsStore.get('autoFetchHours', 24),
    starHistoryLoaded: settingsStore.get('starHistoryLoaded', false),
    showDebugBar: settingsStore.get('showDebugBar', false)
  };
}

function setSettings(partial) {
  if (partial.alwaysOnTop !== undefined) settingsStore.set('alwaysOnTop', !!partial.alwaysOnTop);
  if (partial.startMinimised !== undefined) settingsStore.set('startMinimised', !!partial.startMinimised);
  if (partial.autoFetchHours !== undefined) {
    const h = Number(partial.autoFetchHours);
    if (Number.isFinite(h) && h >= 1) settingsStore.set('autoFetchHours', Math.min(168, h));
  }
  if (partial.starHistoryLoaded !== undefined) settingsStore.set('starHistoryLoaded', !!partial.starHistoryLoaded);
  if (partial.showDebugBar !== undefined) settingsStore.set('showDebugBar', !!partial.showDebugBar);
  return getSettings();
}

function getWindowBounds() {
  return settingsStore.get('windowBounds', null);
}

function setWindowBounds(bounds) {
  settingsStore.set('windowBounds', bounds);
}

function needsAutoFetch(userDataPath) {
  const settings = getSettings();
  const last = settings.lastFetchAt;
  if (!last) return true;
  const hours = settings.autoFetchHours || 24;
  return Date.now() - last > hours * 60 * 60 * 1000;
}

module.exports = {
  loadAllSnapshots,
  saveSnapshot,
  getSettings,
  setSettings,
  getWindowBounds,
  setWindowBounds,
  needsAutoFetch,
  recordTiming,
  getTimingStats,
  settingsStore
};

File: src-tauri/src/store.rs
use std::path::{Path, PathBuf};

use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use tauri_tray_base::{save_settings, TrayBaseState};

use crate::collector::{RepoTotals, Totals};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Snapshot {
    pub timestamp: i64,
    pub date: String,
    pub repos: Vec<RepoTotals>,
    pub totals: Totals,
}

pub fn snapshots_dir(user_data: &Path) -> PathBuf {
    let dir = user_data.join("snapshots");
    let _ = std::fs::create_dir_all(&dir);
    dir
}

pub fn load_all_snapshots(user_data: &Path) -> Vec<Snapshot> {
    let dir = snapshots_dir(user_data);
    let mut files: Vec<_> = std::fs::read_dir(&dir)
        .into_iter()
        .flatten()
        .filter_map(|e| e.ok())
        .filter(|e| e.path().extension().map(|x| x == "json").unwrap_or(false))
        .map(|e| e.path())
        .collect();
    files.sort();
    let mut out = Vec::new();
    for f in files {
        if let Ok(text) = std::fs::read_to_string(&f) {
            if let Ok(s) = serde_json::from_str::<Snapshot>(&text) {
                out.push(s);
            }
        }
    }
    out.sort_by_key(|s| s.timestamp);
    out
}

pub fn save_snapshot(
    user_data: &Path,
    state: &TrayBaseState,
    repos: Vec<RepoTotals>,
    totals: Totals,
) -> Result<Snapshot, String> {
    let snapshot = Snapshot {
        timestamp: chrono::Utc::now().timestamp_millis(),
        date: crate::collector::date_key(),
        repos,
        totals,
    };
    let file = snapshots_dir(user_data).join(format!("{}.json", snapshot.date));
    let text = serde_json::to_string_pretty(&snapshot).map_err(|e| e.to_string())?;
    std::fs::write(file, text).map_err(|e| e.to_string())?;

    {
        let mut settings = state.settings.lock();
        settings
            .extra
            .insert("lastFetchAt".into(), json!(snapshot.timestamp));
        let _ = save_settings(&state.settings_path, &settings);
    }

    Ok(snapshot)
}

pub fn record_timing(state: &TrayBaseState, kind: &str, ms: u64, meta: Value) {
    let mut settings = state.settings.lock();
    let list = settings
        .extra
        .entry("timings")
        .or_insert_with(|| json!([]));
    if let Some(arr) = list.as_array_mut() {
        let mut entry = json!({
            "kind": kind,
            "ms": ms,
            "at": chrono::Utc::now().timestamp_millis()
        });
        if let Some(obj) = meta.as_object() {
            if let Some(e) = entry.as_object_mut() {
                for (k, v) in obj {
                    e.insert(k.clone(), v.clone());
                }
            }
        }
        arr.push(entry);
        while arr.len() > 50 {
            arr.remove(0);
        }
    }
    let _ = save_settings(&state.settings_path, &settings);
}

pub fn timing_stats(state: &TrayBaseState) -> Value {
    let settings = state.settings.lock();
    let list = settings
        .extra
        .get("timings")
        .and_then(|v| v.as_array())
        .cloned()
        .unwrap_or_default();

    fn avg(items: &[&Value]) -> Value {
        if items.is_empty() {
            return Value::Null;
        }
        let sum: f64 = items.iter().filter_map(|t| t.get("ms")?.as_f64()).sum();
        json!((sum / items.len() as f64).round() as u64)
    }

    let auth: Vec<_> = list
        .iter()
        .filter(|t| t.get("kind").and_then(|k| k.as_str()) == Some("auth"))
        .collect();
    let fetch: Vec<_> = list
        .iter()
        .filter(|t| t.get("kind").and_then(|k| k.as_str()) == Some("fetch"))
        .collect();
    let history: Vec<_> = list
        .iter()
        .filter(|t| t.get("kind").and_then(|k| k.as_str()) == Some("history"))
        .collect();

    json!({
        "authAvgMs": avg(&auth),
        "fetchAvgMs": avg(&fetch),
        "historyAvgMs": avg(&history),
        "authCount": auth.len(),
        "fetchCount": fetch.len(),
        "historyCount": history.len(),
        "lastAuth": auth.last().cloned(),
        "lastFetch": fetch.last().cloned(),
        "lastHistory": history.last().cloned()
    })
}
